pkg/healthchecker/healthchecker.go

```diff
@@ -0,0 +1,169 @@
+/*
+Copyright 2021 The Cockroach Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package healthchecker
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/cockroachdb/cockroach-operator/pkg/kube"
+	"github.com/cockroachdb/cockroach-operator/pkg/resource"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"go.uber.org/zap/zapcore"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	underreplicatedmetric = "ranges_underreplicated{store=\"%v\"}"
+	cmdunderreplicted     = "curl -Lk localhost:%s/_status/vars --silent | grep -i '%s'"
+)
+
+type HealthChecker interface { // for testing
+	Probe(ctx context.Context, l logr.Logger, logSuffix string, partition int) error
+}
+
+type HealthCheckerImpl struct {
+	clientset *kubernetes.Clientset
+	scheme    *runtime.Scheme
+	cluster   *resource.Cluster
+	config    *rest.Config
+}
+
+func NewHealthChecker(cluster *resource.Cluster, clientset *kubernetes.Clientset, scheme *runtime.Scheme, config *rest.Config) *HealthCheckerImpl {
+	return &HealthCheckerImpl{
+		clientset: clientset,
+		scheme:    scheme,
+		cluster:   cluster,
+		config:    config,
+	}
+}
+
+func (s *HealthCheckerImpl) Probe(ctx context.Context, l logr.Logger, logSuffix string, nodeID int) error {
+	l.V(int(zapcore.DebugLevel)).Info("Health check probe", "label", logSuffix, "nodeID", nodeID)
+	stsname := s.cluster.StatefulSetName()
+	stsnamespace := s.cluster.Namespace()
+	podname := fmt.Sprintf("%s-%v", stsname, nodeID)
+
+	//we sleep 1 minunte to ensure the pre-stop hook finished
+	time.Sleep(1 * time.Minute)
+	//SKIP KILL because we just restarted????
+	// err := sendKillSignal(ctx, l, logSuffix, podname, nodeID)
+	// if err != nil {
+	// 	return err
+	// }
+
+	//isready check for all pods from sts
+	err := kube.WaitUntilAllStsPodsAreReady(ctx, s.clientset, l, stsname, stsnamespace, 30*time.Minute, 10*time.Minute)
```

Comment:
30 minutes seems excessive. I think after 3m we either fail or move one

Reply:
Also what's the difference between this method and `WaitUntilStatefulSetIsReadyToServe`

Reply:
I added kube.WaitUntilAllStsPodsAreReady in kube package that is common so we use it in all the packages.  Actually WaitUntilStatefulSetIsReadyToServe  will be replace with kube.WaitUntilAllStsPodsAreReady.